File: Dynamic_Programming/901/dp.cpp
#include <iostream>
#include <cstring>
using namespace std;

const int N = 310;
int res, n, m, a, b, h[N][N], dp[N][N], dx[4] = {-1, 1, 0, 0}, dy[4] = {0, 0, -1, 1};

int dfs(int x, int y);

int main() {
    memset(dp, -1, sizeof(dp));
    scanf("%d%d", &n, &m);
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) scanf("%d", &h[i][j]);
    }
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) res = max(res, dfs(i, j));
    }
    printf("%d", res);
    return 0;
}

int dfs(int x, int y) {
    if (dp[x][y] != -1) return dp[x][y];
    dp[x][y] = 1;
    for (int i = 0; i < 4; i++) {
        a = x + dx[i];
        b = y + dy[i];
        if (a >= 1 && a <= n && b >= 1 && b <= m && h[a][b] < h[x][y]) {
            dp[x][y] = max(dp[x][y], dfs(a, b) + 1);
        }
    }
    return dp[x][y];
}

File: Dynamic_Programming/901/dp.py
def dfs(x: int, y: int, h: list, dp: list) -> int:
    if dp[x][y] != -1:
        return dp[x][y]
    dp[x][y] = 1
    dx = (-1, 1, 0, 0)
    dy = (0, 0, -1, 1)
    for i in range(4):
        u, v = x + dx[i], y + dy[i]
        if u >= 0 and u < len(h) and v >= 0 and v < len(h[0]) and h[u][v] < h[x][y]:
            dp[x][y] = max(dp[x][y], dfs(u, v, h, dp) + 1)
    return dp[x][y]

if __name__ == "__main__":
    n, m = map(int, input().split())
    h = [list(map(int, input().split())) for _ in range(n)]
    dp = [[-1] * m for _ in range(n)]
    res = 0
    for i in range(n):
        for j in range(m):
            res = max(res, dfs(i, j, h, dp))
    print(res)
